New single general API endpoint to expose settings based on a new grouping and naming (#10126)
IQSS/dataverse: doc/sphinx-guides/source/api/native-api.rst

  curl "https://demo.dataverse.org/api/info/settings/:MaxEmbargoDurationInMonths"

Get Exposed Settings
~~~~~~~~~~~~~~~~~~~~

> Although the previous endpoints allow to get public settings individually, this endpoint unifies all public settings by exposing them individually or in groups.

Reviewer: ```suggestion
Although the previous endpoints allow retrieval of public settings individually, this endpoint unifies public settings by exposing them individually or in groups.
```

> For supporting this, the public settings have been grouped considering different Dataverse scopes. Within these groups, the settings have been named in a natural way for the user, being this natural name independent of the setting source name on which they depend (JVM options or database settings).

Reviewer: ```suggestion
To support this, the public settings have been grouped. Within these groups, the settings have been named in a natural way for the user, independent of the setting source name on which they depend (JVM options or database settings).
```

Groups follow a tree structure, starting with the root group and its root-level settings, and each child group can contain one or more settings and/or one or more child groups.

Currently under the root group we find the following settings:

- fqdn
- isPublicInstall

And the following child groups, with their associated settings:

- api: termsOfUse, allowIncompleteMetadata.
- dataset: publishPopupCustomText, allowedCurationLabels, zipDownloadLimit.
- datafile: maxEmbargoDurationInMonths.

> The tree-like structure translates to the endpoint structure, which will receive child group names and setting names via path parameters, ordered by depth starting from root.

Reviewer: Is there any performance concern? Could this lead to a denial of service attack?

Author: The root SettingGroup object containing all settings and subgroups is instantiated as a singleton, so reading the settings from all the sources, which is what I think could cause loading issues if done repeatedly, should be done once.

I have added an optimization including thread safety to the singleton instantiation.

For reading a single setting from a child group:

.. code-block:: bash

  export SERVER_URL=https://demo.dataverse.org
  export SETTING_GROUP=api
  export SETTING_NAME=termsOfUse

  curl "$SERVER_URL/api/info/exposedSettings/$SETTING_GROUP/$SETTING_NAME"

The fully expanded example above (without environment variables) looks like this:

.. code-block:: bash

  curl "https://demo.dataverse.org/api/info/exposedSettings/api/termsOfUse"

For reading all settings from a particular group:

.. code-block:: bash

  export SERVER_URL=https://demo.dataverse.org
  export SETTING_GROUP=dataset

  curl "$SERVER_URL/api/info/exposedSettings/$SETTING_GROUP"

The fully expanded example above (without environment variables) looks like this:

.. code-block:: bash

>   curl "https://demo.dataverse.org/api/info/dataset"

Reviewer: ```suggestion
  curl "https://demo.dataverse.org/api/info/exposedSettings/dataset"
```

> For group reading, there are two lookup modes, defined by the optional "lookupMode" query parameter:

Reviewer: ```suggestion
There are two lookup modes, defined by the optional "lookupMode" query parameter:
```

Since we are already using the term "group" above, it might help to remove it (or use a different term) here.

- base: The child groups will not be read and only their names will appear in the response body alongside the parent setting names. This is the default option, if not specified.

> - sub: The child groups will also be read, in order to retrieve all their underlying settings and child groups following the tree structure downwards.

Reviewer: Ok, so to get a dump of all exposed settings, one can do this:

http://localhost:8080/api/info/exposedSettings?lookupMode=sub

Should we mention this somewhere?

Author: Yes, makes sense to mention it

Author: Done
